src/Pipeline/Atomic.php

```diff
@@ -62,56 +64,82 @@ protected function getConnection()
     protected function executePipeline(ConnectionInterface $connection, \SplQueue $commands)
     {
         $profile = $this->getClient()->getProfile();
-        $connection->executeCommand($profile->createCommand('multi'));
+        $multi = $profile->createCommand('MULTI');
+        $exec = $profile->createCommand('EXEC');
 
+        // open transaction: MULTI
+        $connection->writeRequest($multi);
+
+        // transmit all commands in transaction
         foreach ($commands as $command) {
             $connection->writeRequest($command);
         }
 
+        // execute transaction: EXEC
+        $connection->writeRequest($exec);
+
+        // response to MULTI
+        $response = $connection->readResponse($multi);
+        if ($response instanceof ErrorResponseInterface) {
+            $this->exception($connection, $response); // close connection and throw ServerException
+        }
+
+        // responses to commands (all should be QUEUED result)
+        $ex = null;
         foreach ($commands as $command) {
             $response = $connection->readResponse($command);
 
-            if ($response instanceof ErrorResponseInterface) {
-                $connection->executeCommand($profile->createCommand('discard'));
-                throw new ServerException($response->getMessage());
+            // throw first error as a ServerException
+            if (($response instanceof ErrorResponseInterface) && !$ex) {
+                $ex = new ServerException($response->getMessage());
             }
         }
 
-        $executed = $connection->executeCommand($profile->createCommand('exec'));
+        // response to EXEC
+        $responses = $connection->readResponse($exec);
 
-        if (!isset($executed)) {
-            // TODO: should be throwing a more appropriate exception.
-            throw new ClientException(
-                'The underlying transaction has been aborted by the server.'
-            );
+        // if one of the commands returned an error, only throw after all responses have been received
+        if ($ex) {
+            throw $ex;
         }
```

Comment:
I'm most likely missing something here but how is this any different than using [`DISCARD`](https://redis.io/commands/discard) (it tells Redis to discard the queued commands and reset the server-side state without having the client to drop the connection) and throw the exception as soon as we get a non `+QUEUED` response from the server, like did in [previous code](https://github.com/predis/predis/pull/499/files#diff-018d40c6b8462f1b9a5e89720186e531L74-L76)?

And if we use `DISCARD` to clear the transaction there's no actual reason to spend time and resources reading back the responses from the server just to prevent protocol desynchonization issues (just to prove it, if you do a `DISCARD` and then try to do an `EXEC` you'll get a `-ERR EXEC without MULTI`).

Reply:
Alright nevermind, I was indeed missing the fact that we are pipelining here and that's exactly the reason why we had a bug in the first place 🤦‍♂️ I'll leave my original comment above as a reference anyway.

In this context, using [`DISCARD`](https://redis.io/commands/discard) after pipelining all of the writes would be pointless anyway as we have already sent all of the commands to Redis anyway and are just reading back status responses.

That said,  I still think that as soon as we read back an error instead of a `+QUEUED` response it'd be probably more efficient to drop the connection, clear the `SplQueue` object and then throw the exception immediatly instead of consuming all of the outstanding reads. Wouldn't it work the same to avoid protocol desynchronization issues?

Reply:
> That said, and unless I'm wrong once again, I still think that as soon as we read back an error in place of a +QUEUED response it'd be probably more efficient to explicitly drop the connection (which also clears the pending transaction on the Redis side), clear the SplQueue object and then throw the exception immediatly instead of consuming all of the outstanding reads and throw the exception after reading back the response from EXEC. Wouldn't it work the same to avoid protocol desynchronization issues?

@jimbonator do you have any feedback about my observation ⬆️

Reply:
Sorry, I've been caught up in other work and didn't get a chance to come back to this yesterday.

You're right, that would avoid the desynchronization issues, but I was hoping to avoid dropping the connection for server (rather than connection) errors.

Reply:
I've gone ahead and reverted this approach.  When an error is encountered, the connection is dropped and the exception immediately thrown.